show/main.py

```diff
@@ -791,6 +791,21 @@ def mac(vlan, port, verbose):
 
     run_command(cmd, display_cmd=verbose)
 
+#
+# 'show route-map' command ("show route-map")
+#
+
+@cli.command('route-map')
+@click.argument('route_map_name', required=False)
+@click.option('--verbose', is_flag=True, help="Enable verbose output")
+def route_map(route_map_name, verbose):
+    """show route-map"""
+    cmd = 'sudo vtysh -c "show route-map'
+    if route_map_name is not None:
+        cmd += ' {}'.format(route_map_name)
+    cmd += '"'
+    run_command(cmd, display_cmd=verbose)
+	
```

Comment:
Since both commands have options, I would suggest following what I have done for show bgp:

elif routing_stack == "frr":
    @cli.command()
    @click.argument('bgp_args', nargs = -1, required = False)
    @click.option('--verbose', is_flag=True, help="Enable verbose output")
    def bgp(bgp_args, verbose):
        """BGP information"""
        bgp_cmd = "show bgp"
        for arg in bgp_args:
            bgp_cmd += " " + str(arg)
        cmd = 'sudo vtysh -c "{}"'.format(bgp_cmd)
        run_command(cmd, display_cmd=verbose)

That way the command you want will work along with all the options following it.

Reply:
Something like so:
''''
def route_map()route_map_args, verbose):
    route_map_cmd = "show route-map"
    for arg in route_map_args:
        route_map_cmd += " " + str(arg)
    cmd = 'sudo vtysh -c "{}"'.format(route_map_cmd)
    run_command(cmd, display_cmd=verbose)
''''

Reply:
@nikos-github: I like your solution, as it is simple and allows the user to pass multiple parameters. However, without explicitly defining the available parameters in the SONiC CLI, how is the user to know which parameters are available? We can't expect the user to know what is being called under the hood. This is especially a concern with commands like the BGP commands where the underlying implementation is containerized, and can be different on different devices (Quagga, FRR, etc.). I prefer defining all parameters and flags explicitly in the SONiC CLI, that way they appear in the help message for that command.

Reply:
@jleveque The options/parameters can be obtained by "?" as always (doesn't work now but @rodnymolina has it working with vtysh). If LI doesn't want to upstream to FRR I can redevelop and push. Regarding the different routing application, that should be achievable by having the specific .py files depending on the application. We do that today, right? The other advantage of the proposal is that it immediately unlocks the entire parse chain and options without having to redevelop and also shields us from having to react to parse chain changes in the routing application that could cause breakages outside.

Reply:
>The other advantage of the proposal is that it immediately unlocks the entire parse chain and options without having to redevelop and also shields us from having to react to parse chain changes in the routing application that could cause breakages outside.

I completely agree with this. Like I said earlier, I do like this idea ... a lot, actually. However,

>Regarding the different routing application, that should be achievable by having the specific .py files depending on the application. We do that today, right?

Yes, we do. That is not my concern. My concern is that if available options/arguments are not available via the command help, having multiple different underlying applications makes things even worse, as the available options/arguments may be different from platform to platform.

A user should be able to see all available arguments and options for *any and all* SONiC command by simply entering `show <command> --help`. If this is possible with your method, then I'm 100% for using that wherever possible. Can you confirm this?

Reply:
As I mentioned, today that doesn't work because the work in vtysh hasn't been upstreamed. But  it exists and it's working and I can confirm that. Should be no problem to replace --help with "?" and pass it along so that it works too.

Reply:
>Should be no problem to replace --help with "?" and pass it along so that it works too.

The `--help`/`-h`/`-?` flags (whatever we define as a help flag) are processed internally by Click; if any of these are passed, the actual command function is not called; Click outputs a help message and exits. I'm not sure if we can override this behavior.

Reply:
A couple of things:

- There was no problem with "?". We can check for the variations with "-" in front.
- I was actually wrong and the "?" support has already been upstreamed by Rodny (see https://github.com/FRRouting/frr/pull/2221). We will pick it up the moment we upgrade sonic to frr 6.0 (it's in my task list for this month). For the quagga users, we can use the same thing I did for bgp and checking the routing stack. It may not matter after February since the plan is for us to get the default routing stack to be frr in sonic (again, this is in my task list per schedule for this month).

Reply:
> There was no problem with "?". We can check for the variations with "-" in front.

There would be no problem with `?` because it is not a special flag in Click. It would be added to `bgp_args` and passed to vtysh which would produce the help message from vtysh.  However, the `-h/--help/-?` flags are registered as special help flags. These flags are processed by Click. If one of these flags is passed, the underlying command function is not called; Click prints a help message and exits. Again, I'm not sure how easy it is (or if it's even possible) to override behavior for help flags in Click.

I say we merge this PR as-is and wait for the transition to FRR as the default routing stack to implement your solution, hopefully also giving time for someone to research if we can override the behavior for help arguments on a command-by-command basis.

Reply:
I just thought it's a good idea and we could do now. I'm fine either way.